include/pmacc/math/complex/Bessel.tpp

```diff
@@ -0,0 +1,271 @@
+        typedef T_FloatType float_T;
+        T_FloatType const eps = T_FloatType( 1.0e-15 );
+        T_FloatType const el = T_FloatType( 0.5772156649015329 );
+
+        complex_T const cii = complex_T( 0, 1 );
+        complex_T const cone = complex_T( 1, 0 );
+        complex_T const czero = complex_T( 0, 0 );
+
+        HDINLINE result operator( )( complex_T const z )
+        {
+            T_TableA a;
+            T_TableB b;
+            T_TableA1 a1;
+            T_TableB1 b1;
+            result cj1;
+
+            float_T const a0 = pmMath::abs( z );
+            float_T const z2 = z * z;
+            complex_T z1 = z;
+            if ( a0 == float_T( 0.0 ) )
+            {
+                cj1 = czero;
+                return cj1;
+            }
+            if ( z.get_real() < float_T( 0.0 ) ) z1 = -z;
+            if ( a0 <= float_T( 12.0 ) )
+            {
+                cj1 = cone;
+                complex_T cr = cone;
+                for ( uint32_t k = 1 ; k <= 40 ; k++ )
+                {
+                    cr *= float_T( -0.25 ) * z2 / ( k * ( k + float_T( 1.0 ) ) );
+                    cj1 += cr;
+                    if ( pmMath::abs( cr ) < pmMath::abs( cj1 ) * eps) break;
+                }
+                cj1 *= float_T( 0.5 ) * z1;
+            }
+            else {
+                uint32_t kz;
+                if ( a0 >= float_T( 50.0 ) ) kz = 8;      // can be changed to 10
+                else if (a0 >= float_T( 35.0 ) ) kz = 10; //   "      "     "  12
+                else kz = 12;                             //   "      "     "  14
+                complex_T const cu = pmMath::sqrt( float_T( M_2_PI ) / z1 );
+                complex_T const ct2 = z1 - float_T( 0.75 ) * float_T( M_PI );
+                complex_T cp1 = cone;
+                for (uint32_t k=0;k<kz;k++)
+                {
+                    cp1 += a1[k] * pmMath::pow(
+                        z1,
+                        float_T( -2.0 * k - 2.0 )
+                    );
+                }
+                complex_T cq1 = float_T( 0.375 ) / z1;
+                for (uint32_t k = 0 ; k < kz ; k++ )
+                {
+                    cq1 += b1[k] * pmMath::pow(
+                        z1,
+                        float_T( -2.0 * k - 3.0)
+                    );
+                }
+                cj1 = cu * ( cp1 * pmMath::cos( ct2 ) - cq1 * pmMath::sin( ct2 ) );
+            }
+            if ( z.get_real( ) < float_T( 0.0 ) )
+            {
+                cj1 = -cj1;
+            }
+            return cj1;
+        }
+    };
+
+    template< >
+    struct Cbesselj0< double > : public Cbesselj0Base<
+        double,
+        pmacc::algorithms::math::bessel::aDouble_t,
+        pmacc::algorithms::math::bessel::bDouble_t,
+        pmacc::algorithms::math::bessel::a1Double_t,
+        pmacc::algorithms::math::bessel::b1Double_t
+    >{ };
+
+    template< >
+    struct Cbesselj0< float > : public Cbesselj0Base<
```

Comment:
based of #2156 this class must be called `J0`.
Please keep in mind that the name `J0` will be changed or a namespace will be added in #2156 (see [here](https://github.com/ComputationalRadiationPhysics/picongpu/pull/2156#issuecomment-319010061))